Starport blog tutorial and polling app (#349)
cosmos/sdk-tutorials: starport-blog/04-http-create-post.md

# Create posts with HTTP

Let's create a front-end for our blog application. In this guide we will be writing a client-side application in JavaScript that can create a wallet (public/private key pair), fetch a list of posts from our server, create posts and send to our server.

## `x/blog/client/rest/rest.go`

```go
import (
  // Existing imports...
  "github.com/example/blog/x/blog/types"

> 	"github.com/cosmos/cosmos-sdk/x/auth/client/utils"

Reviewer: ```suggestion
  "github.com/cosmos/cosmos-sdk/x/auth/client/utils"
```

Reviewer: wy is this here actually? i don't see it being used.

Author: You're right, I'll remove it.

)
```

We'll be creating posts by sending `POST` requests to the same endpoint: `/blog/posts`. To add a handler add the following line to `func RegisterRoutes`:

```go
	r.HandleFunc("/blog/posts", createPostHandler(cliCtx)).Methods("POST")
```

> Now let's create `createPostHandler`.

Reviewer: ```suggestion
Now let's create `createPostHandler` inside a new file you should create at `x/blog/client/rest/tx.go`.
```

## `x/blog/client/rest/tx.go`

We will need a `createPostReq` type that represents the request that we will be sending from the client:

```go
package rest

import (
	"net/http"

	"github.com/cosmos/cosmos-sdk/client/context"
	sdk "github.com/cosmos/cosmos-sdk/types"
	"github.com/cosmos/cosmos-sdk/types/rest"
	"github.com/cosmos/cosmos-sdk/x/auth/client/utils"
	"github.com/example/blog/x/blog/types"
)

type createPostReq struct {
	BaseReq rest.BaseReq `json:"base_req"`
	Creator string       `json:"creator"`
	Title   string       `json:"title"`
}
```

> `createPostHandler` first parses request parameters, performs basic validations, converts `Creator` field from string into SDK account address type, creates `MsgCreatePost` message.

Reviewer: ```suggestion
`createPostHandler` first parses request parameters, performs basic validations, converts `Creator` field from a string into an SDK account address type then finally creates `MsgCreatePost` message.
```

```go
func createPostHandler(cliCtx context.CLIContext) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var req createPostReq
		if !rest.ReadRESTReq(w, r, cliCtx.Codec, &req) {
			rest.WriteErrorResponse(w, http.StatusBadRequest, "failed to parse request")
			return
		}
		baseReq := req.BaseReq.Sanitize()
		if !baseReq.ValidateBasic(w) {
			return
		}
		addr, err := sdk.AccAddressFromBech32(req.Creator)
		if err != nil {
			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
			return
		}
		msg := types.NewMsgCreatePost(addr, req.Title)
		err = msg.ValidateBasic()
		if err != nil {
			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
			return
		}
		utils.WriteGenerateStdTxResponse(w, cliCtx, baseReq, []sdk.Msg{msg})
	}
}
```

## Setting up the client-side project

> Inside an empty directory create `package.json`:

Reviewer: ```suggestion
Inside a new empty directory create `package.json`:
```

```sh
npm init -y
```

Add dependencies:

```sh
npm add parcel-bundler local-cors-proxy axios @tendermint/sig
```

We'll be using `parcel-bundler` for bundling our dependencies and development server, `local-cors-proxy` for providing a CORS proxy server as a development replacement for something like Nginx, `axios` for HTTP requests and `@tendermint/sig` for interacting with our application.

Replace `"scripts"` property in `package.json` with the following:

```json
"scripts": {
  "preserve": "lcp --proxyUrl http://localhost:1317 &",
  "serve": "parcel index.html",
  "postserve": "pkill -f lcp"
}
```

> Responses from an HTTP server provided by `blogcli rest-server` do not provide `Access-Control-Allow-Origin` headers. This prevents client-side application running from different domains/ports from making successful requests due to browser's CORS policy. In production that can be managed by a web server like Nginx providing the right headers. In development we'll be using `local-cors-proxy` (`lcp`) to act as a proxy between our front-end and server.

Reviewer: ```suggestion
Responses from an HTTP server provided by `blogcli rest-server` do not provide `Access-Control-Allow-Origin` headers. This prevents client-side application running from different domains/ports from making successful requests due to most browsers' CORS policies. In production that can be managed by a web server like Nginx providing the right headers. In development we'll be using `local-cors-proxy` (`lcp`) to act as a proxy between our front-end and server.
```

> `preserve` runs first and launches `lcp` in background. `serve` launches a development server and `postserve` kills `lcp` process after you shutdown the server.

Reviewer: ```suggestion
The commands under `scripts` work as follows: `preserve` runs first and launches `lcp` in background. `serve` launches a development server and `postserve` kills `lcp` process after you shutdown the server.
```

## `index.html`

Our `index.html` file will contain only the minimal amount of markup:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>Blog</title>
  </head>
  <body>
    <h1>Blog</h1>
    <input placeholder="Post title" type="text" id="input" />
    <button id="button">Create post</button>
    <div id="posts"></div>
    <script src="script.js"></script>
  </body>
</html>
```

`<input/>` provides a text input for a title of a new post. `<button>` will submit a new post to our server. `<div id="post>` is an empty container which will be populated by a list of posts. We load `script.js` after all other markup has been loaded.

Now let's create `script.js`.

## `script.js`

Our app will depend on `axios` and `@tendermint/sig`:

```js
import axios from "axios";
import {
  createWalletFromMnemonic,
  signTx,
  createBroadcastTx,
} from "@tendermint/sig";
```

> 

Reviewer: why a gap here?

```js
const API = "http://localhost:8010/proxy";
```

HTTP-server runs on `localhost:1317`. Due to CORS we need to route our requests through a proxy: we're using `local-cors-proxy`, which uses the above URL by default.

First, we need to create an account: a public/private key pair, which we will use for signing our (create post) transactions before broadcasting them to the server. We'll be using a mnemonic, from which the keys will be generated.

```js
const mnemonic =
  "solid play vibrant paper clinic talent people employ august camp april reduce";
```

Now let's generate a wallet:

```js
const wallet = createWalletFromMnemonic(mnemonic);
```

> If you use the same mnemonic (think of it as a password), you will get the same wallet address: in out case `wallet.address` is `cosmos152gzu3vzf7g9tu46vszgpac24lwr48vc8k8kkh`. If you want to generate unique mnemonics for your users, you can use `bip39` package.

Reviewer: link bip39?

Our app will do two things: it will fetch a list of posts and create posts when "Create post" button is clicked.

```js
const init = () => {
  fetchPosts();
  document.getElementById("button").addEventListener("click", createPost);
};
```

For `init()` to work we need to define two functions: `fetchPosts` and `createPost`.

```js
const fetchPosts = () => {
  axios.get(`${API}/blog/posts`).then(({ data }) => {
    const posts = JSON.stringify(data.result);
    document.getElementById("posts").innerText = posts;
  });
};
```

`fetchPosts` makes an HTTP GET request to `/blog/posts` and inserts the posts into the `<div id="posts">` container.

> Next, we need to define `createPost` function. It will take the value from the text input, fetch required account parameters from the server, fetch an unsigned transaction from the server, sign it using our private key from the wallet, broadcast it back to the server and fetch the list of posts again.

Reviewer: ```suggestion
Next, we need to define the `createPost` function. It will take the value from the text input, fetch the required account parameters from the server, fetch an unsigned transaction from the server, sign it using our private key from the wallet, broadcast it back to the server and fetch the list of posts again.
```

      // the transaction is included in the block
      const txBroadcast = createBroadcastTx(stdTx, "block");
      const params = {
        headers: {
          "Content-Type": "application/json",
        },
      };
      // Sending our post to be processed by the server
      axios.post(`${API}/txs`, txBroadcast, params).then(() => {
        // Fetch a new list of posts after we've successfully
        // created a new post
        fetchPosts();
      });
    });
  });
};
```

Run init to initialize our app:

```js
init();
```

## Creating an account

Before we start using our application we need to make sure that the account we have generated exists on our chain. To do so we will send a nominal amount of tokens from an existing account to a new one.

```sh
blogcli keys show user1
```

You will get information about one of the existing accounts (your values will be different):

```json
{
  "name": "user1",
  "type": "local",
  "address": "cosmos1wt47yve6l29yjtxtsajhltr2vqhf7mpw5n6fx6",
  "pubkey": "cosmospub1addwnpepq03v7d6q4yt4nalj74elq8l5498wd9krcx92mxudkarj8aapy0qjvfaga8z"
}
```

Transfer some tokens from this account to the new one:

```sh

> blogcli tx send cosmos1wt47yve6l29yjtxtsajhltr2vqhf7mpw5n6fx6 cosmos152gzu3vzf7g9tu46vszgpac24lwr48vc8k8kkh 10foo --from=user1

Reviewer: ```suggestion
blogcli tx send $(blogcli keys show user1 -a) cosmos152gzu3vzf7g9tu46vszgpac24lwr48vc8k8kkh 10foo --from=user1
```

```

> Notice that the sender address `cosmos1wt47yve6l29yjtxtsajhltr2vqhf7mpw5n6fx6` is the one we got from running `blogcli keys show me` and the receiver account address `cosmos152gzu3vzf7g9tu46vszgpac24lwr48vc8k8kkh` is the one we have generated from the mnemonic in browser.

Reviewer: ```suggestion
Notice that the sender address can be queried automatically using the sub-command `$(blogcli keys show user1 -a)` with the flag `-a` to show just the address and the receiver account address `cosmos152gzu3vzf7g9tu46vszgpac24lwr48vc8k8kkh` is the one we have generated from the mnemonic in the browser.
```

In this guide we're activating accounts manually, but in production apps you might want to do it as part of a signing up process.

> Now that you that we've created the app and set up our account, let's run it!

Reviewer: ```suggestion
Now that we've created the app and set up our account, let's run it!
```

```
npm run serve
```

> Open `http://localhost:1234/` and try creating new posts. New posts should show up in the list after 2-4 seconds after submission. Notice that posts persist after you refresh the page.

Reviewer: I get this error in the browser console:
```
xhr.js:178 POST http://localhost:8010/proxy/blog/posts 405 (Method Not Allowed)
```

Author: Seems like proxy didn't work. I've created a repo with branches for each step. Checked multiple times, couldn't reproduce this error.

https://github.com/fadeev/tutorial-blog/tree/04-http-create-posts

Author: There was one more error: `foo` instead of `token`. Starport used to init with `foo` (like Scavenge), but we have replaced it with `token`.
